pkg/protocol/block_dispatcher.go

```diff
@@ -0,0 +1,345 @@
+	}, opts, func(b *BlockDispatcher) {
+		protocol.HookConstructed(b.initEngineMonitoring)
+		protocol.HookInitialized(b.initNetworkConnection)
+		protocol.HookStopped(b.shutdown)
+	})
+}
+
+// Dispatch dispatches the given block to the correct engine instance.
+func (b *BlockDispatcher) Dispatch(block *model.Block, src network.PeerID) error {
+	slotCommitment := b.protocol.ChainManager.LoadCommitmentOrRequestMissing(block.ProtocolBlock().SlotCommitmentID)
+	if !slotCommitment.SolidEvent().WasTriggered() {
+		if !b.unsolidCommitmentBlocks.Add(slotCommitment.ID(), types.NewTuple(block, src)) {
+			return ierrors.Errorf("failed to add block %s to unsolid commitment buffer", block.ID())
+		}
+
+		return ierrors.Errorf("failed to dispatch block %s: slot commitment %s is not solid", block.ID(), slotCommitment.ID())
+	}
+
+	matchingEngineFound := false
+	for _, engine := range []*engine.Engine{b.protocol.MainEngineInstance(), b.protocol.CandidateEngineInstance()} {
+		if engine != nil && (engine.ChainID() == slotCommitment.Chain().ForkingPoint.ID() || engine.BlockRequester.HasTicker(block.ID())) {
+			if !b.inWarpSyncRange(engine, block) {
+				engine.ProcessBlockFromPeer(block, src)
+			}
+
+			matchingEngineFound = true
+		}
+	}
+
+	if !matchingEngineFound {
+		return ierrors.Errorf("failed to dispatch block %s: no matching engine found", block.ID())
+	}
+
+	return nil
+}
+
+// initEngineMonitoring initializes the automatic monitoring of the engine instances.
+func (b *BlockDispatcher) initEngineMonitoring() {
+	b.monitorLatestEngineCommitment(b.protocol.mainEngine)
+
+	b.protocol.engineManager.OnEngineCreated(b.monitorLatestEngineCommitment)
+
+	b.protocol.Events.ChainManager.CommitmentPublished.Hook(func(chainCommitment *chainmanager.ChainCommitment) {
+		chainCommitment.SolidEvent().OnTrigger(func() {
+			b.runTask(func() {
+				b.injectUnsolidCommitmentBlocks(chainCommitment.Commitment().ID())
+			}, b.dispatchWorkers)
+
+			b.runTask(func() {
+				// warpsync only if the observed commitment is at least two commitments ahead.
+				targetEngine := b.targetEngine(chainCommitment)
+				if targetEngine != nil && chainCommitment.Commitment().Index() > targetEngine.Storage.Settings().LatestCommitment().Index()+1 {
+					b.warpSync(targetEngine, chainCommitment)
+				}
+			}, b.warpSyncWorkers)
+		})
+	})
+
+	b.protocol.Events.Engine.Notarization.LatestCommitmentUpdated.Hook(func(commitment *model.Commitment) {
+		b.runTask(func() {
+			b.injectUnsolidCommitmentBlocks(commitment.ID())
+		}, b.dispatchWorkers)
+	})
+
+	b.protocol.Events.Engine.SlotGadget.SlotFinalized.Hook(b.evict)
+}
+
+// initNetworkConnection initializes the network connection of the BlockDispatcher instance.
+func (b *BlockDispatcher) initNetworkConnection() {
+	b.protocol.Events.Engine.BlockRequester.Tick.Hook(func(blockID iotago.BlockID) {
+		b.runTask(func() {
+			b.protocol.networkProtocol.RequestBlock(blockID)
+		}, b.dispatchWorkers)
```

Comment:
Why using `b.dispatchWorkers`? Oh I see, seems that otherwise there's a task conention or so because `processWarpSyncResponse` triggers `BlockMissing` which in turn starts the ticker and then all these tasks would be queue until 
```
for _, blockID := range blockIDs {
	targetEngine.BlockDAG.GetOrRequestBlock(blockID)
}
```

is done